AmberScript Attach transcription fails if video contains no speech

If you process a video file which does not contain any speech like [goat.mp4](https://data.lkiesow.io/opencast/test-media/goat.mp4), the attach operation will fail with a rather strange error:

```
2023-08-30T15:00:47,657 | INFO  | (AmberscriptTranscriptionService$WorkflowDispatcher:1090) - Attach transcription workflow 228272 scheduled for mp 0fbf4fb3-1dd6-4a8b-a58c-c93a6588a47a, transcription service job 64ef3bfc846511450ce2e120
2023-08-30T15:00:49,373 | INFO  | (Log:213) - [>559a0c0c] Scheduling workflow 228272 for execution
2023-08-30T15:00:51,403 | INFO  | (AmberscriptTranscriptionService:846) - Looking for transcript at URI: https://testing-oc-admin.uni-osnabrueck.de/files/collection/amberscript-transcripts/64ef3bfc846511450ce2e120.srt
2023-08-30T15:00:51,410 | WARN  | (WorkspaceImpl:386) - The working file repository and workspace paths don't match. Looking up https://testing-oc-admin.uni-osnabrueck.de/files/collection/amberscript-transcripts/64ef3bfc846511450ce2e120.srt at /mnt/opencaststorage/oc-testing/files/collection/amberscript-transcripts/64ef3bfc846511450ce2e120.srt failed
2023-08-30T15:00:51,451 | INFO  | (AmberscriptTranscriptionService:852) - Results not saved: getting from service for jobId 64ef3bfc846511450ce2e120
2023-08-30T15:00:51,709 | WARN  | (AmberscriptTranscriptionService:739) - Captions job '64ef3bfc846511450ce2e120' failed.
2023-08-30T15:00:51,710 | ERROR | (WorkflowOperationWorker:140) - Workflow operation 'operation:'amberscript-attach-transcription, state:'FAILED'' failed
org.opencastproject.workflow.api.WorkflowOperationException: org.opencastproject.transcription.api.TranscriptionServiceException: Captions job '64ef3bfc846511450ce2e120' failed: Return Code 200
        at org.opencastproject.transcription.workflowoperation.AmberscriptAttachTranscriptionOperationHandler.start(AmberscriptAttachTranscriptionOperationHandler.java:135) ~[?:?]
        at org.opencastproject.workflow.impl.WorkflowOperationWorker.start(WorkflowOperationWorker.java:212) ~[?:?]
        at org.opencastproject.workflow.impl.WorkflowOperationWorker.execute(WorkflowOperationWorker.java:117) ~[?:?]
        at org.opencastproject.workflow.impl.WorkflowServiceImpl.runWorkflowOperation(WorkflowServiceImpl.java:723) ~[?:?]
        at org.opencastproject.workflow.impl.WorkflowServiceImpl.process(WorkflowServiceImpl.java:1741) ~[?:?]
        at org.opencastproject.workflow.impl.WorkflowServiceImpl$JobRunner.call(WorkflowServiceImpl.java:2102) ~[?:?]
        at org.opencastproject.workflow.impl.WorkflowServiceImpl$JobRunner.call(WorkflowServiceImpl.java:2068) ~[?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
        at java.lang.Thread.run(Thread.java:829) ~[?:?]
Caused by: org.opencastproject.transcription.api.TranscriptionServiceException: Captions job '64ef3bfc846511450ce2e120' failed: Return Code 200
        at org.opencastproject.transcription.amberscript.AmberscriptTranscriptionService.checkJobResults(AmberscriptTranscriptionService.java:741) ~[?:?]
        at org.opencastproject.transcription.amberscript.AmberscriptTranscriptionService.getGeneratedTranscription(AmberscriptTranscriptionService.java:854) ~[?:?]
        at org.opencastproject.transcription.workflowoperation.AmberscriptAttachTranscriptionOperationHandler.start(AmberscriptAttachTranscriptionOperationHandler.java:113) ~[?:?]
        ... 10 more
```

This fails with `Return Code 200`? That's weird with 200 being a success code.

---

To reproduce this error, take an Opencast 14 and modify the `fast.yml` qorkflow by adding before the `cleanup` operation:

```yml
- id: encode                                                                                                                
  fail-on-error: true                                                                                                       
  exception-handler-workflow: partial-error                                                                              
  description: Encoding audio for transcription                                                                          
  configurations:                                                                                                        
    - source-flavor: "*/source"                                                                                          
    - target-flavor: audio/mp3                                                                                           
    - target-tags: transcript                                                                                            
    - encoding-profile: audio-mp3                                                                                        
                                                                                                                         
- id: amberscript-start-transcription                                                                                    
  max-attempts: 3                                                                                                           
  retry-strategy: hold                                                                                                   
  fail-on-error: false                                                                                                   
  exception-handler-workflow: partial-error                                                                              
  description: Start AmberScript transcription job                                                                       
  configurations:                                                                                                        
    - source-tag: transcript                                                                                             
    - language: de                                                                                                       
    - jobtype: direct                                                                                                    
    - skip-if-flavor-exists: captions/vtt
```

Then add the attach workflow in `uos-amberscript-attach-transcription.yaml`:

```yml
---
id: uos-amberscript-attach-transcription
title: Attach caption/transcripts generated by AmberScript
description: Attach transcription generated by the AmberScript service.
             This is an internal workflow, started by the Transcription Service.

operations:

  - id: amberscript-attach-transcription
    fail-on-error: true
    exception-handler-workflow: partial-error
    description: Attach captions/transcription
    configurations:
      - transcription-job-id: ${transcriptionJobId}
      - target-caption-format: vtt
      - target-tag: engage-download

  - id: publish-engage
    fail-on-error: true
    exception-handler-workflow: partial-error
    description: Distribute and publish to engage server
    configurations:
      - download-source-flavors: "dublincore/*,security/*"
      - download-source-tags: engage-download
      - strategy: merge
      - check-availability: false

  - id: snapshot
    fail-on-error: true
    exception-handler-workflow: partial-error
    description: Archive media package
    configurations:
      - source-flavors: "*/*"

  - id: cleanup
    fail-on-error: false
    description: Remove temporary processing artifacts
    configurations:
      - delete-external: false
      - preserve-flavors: "security/*"
```

Then make sure to update the `org.opencastproject.transcription.amberscript.AmberscriptTranscriptionService.cfg` configuration file:

```properties
# Set true to enable AmberScript Transcription Service
# Default: false
enabled=true

# AmberScript API key
client.key=*******************

# Default language to be used transcribing.
# You can override this within you workflow (see workflow operation handler `amberscript-start-transcription`).
# Can be one of: nl, en, de, fr, da, sv, fi, no, es
# Default: en
language=de

# Set to true if the transcription language should be taken from the DublinCore catalog.
# If the DublinCore catalog doesn't contain a language, the default language will be used.
# Note, that the language field in the workflow operation will be overwritten,
# in case the DublinCore catalog has a language.
# Default: false
#language.from.dublincore=false

# Manual mapping for language codes.
# Here you can set which language spellings/codes from your system shall be
# mapped to the corresponding Amberscript language. In general the Amberscript workflow
# should figure it out by itself. Use this, if you have some special language strings and run into errors.
# You can map a language to the Amberscript code with a colon ':'.
# You can separate multiple mappings with a comma ','.
# Examples:
# language.code.map=english:en
# language.code.map=german:de,french:fr,spanish:es
# language.code.map=SomeWeirdL4nguageC0de:en
# These are three different examples, please note that you can specify 'language.code.map' only once
#language.code.map=

# Default job type to use.
# You can override this within you workflow (see workflow operation handler `amberscript-start-transcription`).
# Can be one of: perfect, direct
# Default: direct
# jobtype=direct

# Workflow to be executed when results are ready to be attached to media package (workflow-id)
# Default: amberscript-attach-transcripts
workflow=uos-amberscript-attach-transcription

# Interval the workflow dispatcher runs to start workflows to attach transcripts to the media package
# after the transcription job is completed in seconds.
# Default: 1 minute
#workflow.dispatch.interval=360

# How long to wait after a transcription is supposed to finish before marking the job as
# cancelled in the database in seconds
# Default: 48 hours
#max.overdue.time=691200

# How long to keep result files in the working file repository in days.
# Default: 7 days
#cleanup.results.days=7

# Default number of speakers to assume.
# You can override this within you workflow (see workflow operation handler `amberscript-start-transcription`).
# Default: 1
#speaker=1

# Set to true if the number of speakers should be taken from the DublinCore catalog.
# If the number of speakers in the DublinCore catalog specified below is 0,
# the default will be used. Note, that the speaker field in the workflow operation
# will be overwritten, in case there are speakers in the DublinCore catalog.
# Default=true
#speaker.from.dublincore=true

# Configure which metadata field shall be considered when "counting" the
# number of speakers. The "presenter" field, the "contributor" field
# or both (i.e. the union of those lists).
# Note for developers: the "presenter" becomes the "creator" in the Operation Handler
# possible values: presenter, contributor, both
# Default: presenter
#speaker.metadata.field=presenter

# Default transcription type to use.
# You can override this within you workflow (see workflow operation handler `amberscript-start-transcription`).
# Can be one of: transcription, captions, translatedSubtitles
# Default: transcription
#transcriptiontype=transcription

# Default glossary to use.
# You can override this within you workflow (see workflow operation handler `amberscript-start-transcription`).
# By default, no glossary is used.
#glossary=

# Default transcription style to use.
# You can override this within you workflow (see workflow operation handler `amberscript-start-transcription`).
# Can be one of: cleanread, verbatim
# Default: cleanread
#transcriptionstyle=cleanread

# The default target language of the transcriptions.
# You can override this within you workflow (see workflow operation handler `amberscript-start-transcription`).
# Can be one of: nl, en, de, fr, da, sv, fi, no, es
#targetlanguage=
```

And finally try processing the [goat.mp4](https://data.lkiesow.io/opencast/test-media/goat.mp4) test media using the fast testing workflow

--- a/modules/transcription-service-amberscript/src/main/java/org/opencastproject/transcription/amberscript/AmberscriptTranscriptionService.java
+++ b/modules/transcription-service-amberscript/src/main/java/org/opencastproject/transcription/amberscript/AmberscriptTranscriptionService.java
@@ -132,6 +132,8 @@ public class AmberscriptTranscriptionService extends AbstractJobProducer impleme
   private static final String STATUS_DONE = "DONE";
   private static final String STATUS_ERROR = "ERROR";
 
+  private static final String ERROR_NO_SPEECH = "No speech found";
+
   private static final String PROVIDER = "amberscript";
 
   private AssetManager assetManager;
@@ -736,9 +738,11 @@ public class AmberscriptTranscriptionService extends AbstractJobProducer impleme
               logger.debug("Captions job '{}' has not finished yet.", jobId);
               return false;
             case STATUS_ERROR:
-              logger.warn("Captions job '{}' failed.", jobId);
+              var errorMsg = (String) result.get("errorMsg");
               throw new TranscriptionServiceException(
-                      String.format("Captions job '%s' failed: Return Code %d", jobId, code), code);
+                      String.format("Captions job '%s' failed: %s", jobId, errorMsg),
+                      code,
+                      ERROR_NO_SPEECH.equals(errorMsg));
             case STATUS_DONE:
               logger.info("Captions job '{}' has finished.", jobId);
               TranscriptionJobControl jc = database.findByJob(jobId);

--- a/modules/transcription-service-api/src/main/java/org/opencastproject/transcription/api/TranscriptionServiceException.java
+++ b/modules/transcription-service-api/src/main/java/org/opencastproject/transcription/api/TranscriptionServiceException.java
@@ -24,6 +24,7 @@ public class TranscriptionServiceException extends Exception {
   private static final long serialVersionUID = 4196196907868554450L;
 
   private int code;
+  private boolean cancel = false;
 
   public TranscriptionServiceException() {
     super();
@@ -42,7 +43,17 @@ public class TranscriptionServiceException extends Exception {
     this.code = code;
   }
 
+  public TranscriptionServiceException(String message, int code, boolean cancel) {
+    super(message);
+    this.code = code;
+    this.cancel = cancel;
+  }
+
   public int getCode() {
     return this.code;
+  }
+
+  public boolean isCancel() {
+    return cancel;
   }
 }

--- a/modules/transcription-service-workflowoperation/src/main/java/org/opencastproject/transcription/workflowoperation/AmberscriptAttachTranscriptionOperationHandler.java
+++ b/modules/transcription-service-workflowoperation/src/main/java/org/opencastproject/transcription/workflowoperation/AmberscriptAttachTranscriptionOperationHandler.java
@@ -29,6 +29,7 @@ import org.opencastproject.mediapackage.MediaPackageElementFlavor;
 import org.opencastproject.mediapackage.MediaPackageElementParser;
 import org.opencastproject.serviceregistry.api.ServiceRegistry;
 import org.opencastproject.transcription.api.TranscriptionService;
+import org.opencastproject.transcription.api.TranscriptionServiceException;
 import org.opencastproject.workflow.api.AbstractWorkflowOperationHandler;
 import org.opencastproject.workflow.api.ConfiguredTagsAndFlavors;
 import org.opencastproject.workflow.api.WorkflowInstance;
@@ -128,9 +129,14 @@ public class AmberscriptAttachTranscriptionOperationHandler extends AbstractWork
         convertedTranscription.addTag(tag);
       }
       mediaPackage.add(convertedTranscription);
-      logger.info("Added transcription to the mediapackage {}: {}",
-          mediaPackage.getIdentifier(), convertedTranscription.getURI());
+      logger.info("Added transcription to the media package {}: {}", mediaPackage, convertedTranscription.getURI());
 
+    } catch (TranscriptionServiceException e) {
+      if (e.isCancel()) {
+        logger.warn(e.getMessage());
+        return createResult(mediaPackage, Action.SKIP);
+      }
+      throw new WorkflowOperationException(e);
     } catch (Exception e) {
       throw new WorkflowOperationException(e);
     }

